torch/csrc/autograd/record_function.cpp

```diff
@@ -233,7 +240,19 @@ void RecordFunction::end() {
         ": must be top of stack. If you are calling RecordFunction::end in a"
         "separate thread, call RecordFunction::setThreadId() in the creating"
         "thread.");
-    thread_local_func_ = parent_;
+
+    // For nested RecordFunction scopes to work when calling
+    // RecordFunction::end() from a different thread, we access a pointer to the
+    // thread_local_func belonging to the thread that started this
+    // RecordFunction.
+    if (threadId_ != 0) {
+      {
+        std::lock_guard<std::mutex> guard(original_mutex);
+        *original_thread_local_func = parent_;
```

Comment:
This is sort of a weird approach where another thread is modifying a different thread's `thread_local` memory. Though, I'm not sure what a better fix might be here. We need to always set `thread_local_func = parent_` when ending the current `RecordFunction` so that nested scopes are handled properly.